src/core/interpreter/map-interpreter.ts

```diff
@@ -85,770 +72,1318 @@ export interface MapParameters<
-  private stackTop(assertType?: 'map' | 'operation'): Stack {
-    if (this.stack.length === 0) {
-      throw new UnexpectedError('Trying to get variables out of scope!');
+  // eslint-disable-next-line require-yield
+  public async *performStream(
+    ast: MapDocumentNode
+  ): AsyncGenerator<
+    unknown,
+    Result<
+      Variables | undefined,
+      MapInterpreterError | UnexpectedError | SDKExecutionError
+    >,
+    undefined
+  > {
+    if (!isMapDocumentNode(ast)) {
+      return err(new UnexpectedError('Invalid AST'));
     }
-    const stack = this.stack[this.stack.length - 1];
 
-    if (assertType !== undefined && stack.type !== assertType) {
-      throw new UnexpectedError(
-        `Trying to get '${assertType}', but got ${stack.type}!`
+    // setup
+    const operations: Record<string, OperationDefinitionNode | undefined> =
+      Object.fromEntries(
+        ast.definitions
+          .filter(isOperationDefinitionNode)
+          .map(op => [op.name, op])
+      );
+    const entry = ast.definitions
+      .filter(isMapDefinitionNode)
+      .find(definition => definition.usecaseName === this.parameters.usecase);
+    if (entry === undefined) {
+      return err(
+        new MapASTError(
+          `Usecase not found: ${this.parameters.usecase ?? 'undefined'}!`,
+          { node: ast, ast }
+        )
       );
     }
 
-    return stack;
+    // create a visitor of the root node and put it on the stack
+    const nodeStack: NodeVisitor<MapASTNode, unknown>[] = [
+      this.createVisitor(
+        entry,
+        {
+          input: this.parameters.input ?? {},
+          parameters: this.parameters.parameters ?? {},
+        },
+        'root'
+      ),
+    ];
+
+    // drive nodes from the stack until empty
+    let lastResult: VisitorResultDone | undefined = undefined;
+    while (nodeStack.length > 0) {
+      const current = nodeStack[nodeStack.length - 1];
+
+      this.log?.('Stepping', current.toString(), '<<', lastResult);
+      // this cast is weird but there is no way to cast to `void | VisitorResultDone<unknown>`
+      const step: VisitorIteratorResult<unknown> = await current.next(
+        lastResult as unknown as VisitorResultDone<unknown>
+      );
+      this.log?.(
+        step.done === true ? 'Returned' : 'Yielded',
+        current.toString(),
+        '>>',
+        step.value
+      );
+
+      lastResult = undefined;
+      switch (step.value.kind) {
+        case 'explore':
+          {
+            let node;
+            if ('node' in step.value.what) {
+              node = step.value.what.node;
+            } else {
+              node = operations[step.value.what.operation];
+              if (node === undefined) {
+                return err(
+                  new MapASTError(
+                    `Operation not found: ${step.value.what.operation}`,
+                    { node: current.node, ast }
+                  )
+                );
+              }
+            }
```

Comment:
question: handling operation calls seems to be disconnected from rest of AST walking. Is it because we need to jump in AST tree to operation?

Reply:
Yes, the options I considered were to either pass `operations` record into the visitor or to have the visitor communicate that it wants to call an operation.

I opted for the second option as it centralizes error handling if the operation is not found and keeps the convention that visitors do not have access to the entire AST but rather only the relevant node.